File: src/test/kotlin/com/marcel/baumgardt/controller/MovieControllerIntegrationTest.java
package com.marcel.baumgardt.controller;

import com.marcel.baumgardt.IntegrationTest;
import com.marcel.baumgardt.TestUtils;
import com.marcel.baumgardt.model.dto.response.MovieDetailResponse;
import lombok.RequiredArgsConstructor;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.test.context.support.WithMockUser;
import org.springframework.test.web.servlet.MvcResult;

import static com.marcel.baumgardt.TestConstants.GET_MOVIE_DETAIL_URL;
import static com.marcel.baumgardt.TestConstants.SUCCESSFUL_MOVIE_ID;
import static org.junit.jupiter.api.Assertions.assertEquals;

@RequiredArgsConstructor(onConstructor = @__(@Autowired))
public class MovieControllerIntegrationTest extends IntegrationTest {

    @WithMockUser(username = "user", authorities = {"USER"})
    @Test
    void shouldGetSuccessfulMovieDetailsResponseAsUser() throws Exception {
        MovieDetailResponse expected = TestUtils.Companion.createSuccessfulMovieDetailResponse();

        MvcResult mvcResult = performGetRequest(GET_MOVIE_DETAIL_URL, SUCCESSFUL_MOVIE_ID);
        MovieDetailResponse actual = getResponseObject(mvcResult, MovieDetailResponse.class);

        assertEquals(actual, expected);
    }

    @WithMockUser(username = "admin", authorities = {"ADMIN"})
    @Test
    void shouldGetSuccessfulMovieDetailsResponseAsAdmin() throws Exception {
        MovieDetailResponse expected = TestUtils.Companion.createSuccessfulMovieDetailResponse();

        MvcResult mvcResult = performGetRequest(GET_MOVIE_DETAIL_URL, SUCCESSFUL_MOVIE_ID);
        MovieDetailResponse actual = getResponseObject(mvcResult, MovieDetailResponse.class);

        assertEquals(actual, expected);
    }

    @Test
    void shouldThrowUnauthorizedExceptionWhenGetMovieDetails() {
        //TODO
    }

    @Test
    void shouldNotFoundMovieDetails() {
        //TODO
    }
}

File: src/test/kotlin/com/marcel/baumgardt/controller/MovieControllerIntegrationTest.kt
package com.marcel.baumgardt.controller

import com.marcel.baumgardt.IntegrationTest
import com.marcel.baumgardt.TestConstants.Companion.ERROR_MOVIE_ID
import com.marcel.baumgardt.TestConstants.Companion.GET_MOVIE_DETAIL_URL
import com.marcel.baumgardt.TestConstants.Companion.SUCCESSFUL_MOVIE_ID
import com.marcel.baumgardt.TestUtils.Companion.createNotFoundMovieDetailResponse
import com.marcel.baumgardt.TestUtils.Companion.createSuccessfulMovieDetailResponse
import com.marcel.baumgardt.model.dto.response.MovieDetailResponse
import lombok.RequiredArgsConstructor
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test

@RequiredArgsConstructor
internal abstract class MovieControllerIntegrationTest : IntegrationTest() {

    @Test
    fun shouldGetSuccessfulMovieDetails() {
        val expected: MovieDetailResponse = createSuccessfulMovieDetailResponse()

        val mvcResult = performGetRequest(GET_MOVIE_DETAIL_URL, SUCCESSFUL_MOVIE_ID)
        val actual: MovieDetailResponse = getResponseObject(mvcResult, MovieDetailResponse::class.java)

        assertEquals(expected, actual)
    }

    @Test
    fun shouldGetNotFoundGetMovieDetails() {
        val expected: MovieDetailResponse = createNotFoundMovieDetailResponse()

        val mvcResult = performGetRequest(GET_MOVIE_DETAIL_URL, ERROR_MOVIE_ID)
        val actual: MovieDetailResponse = getResponseObject(mvcResult, MovieDetailResponse::class.java)

        assertEquals(expected, actual)
    }
}